src/dataflow/operator.rs

```diff
@@ -0,0 +1,310 @@
+                logic(&mut input, &mut ok_output_handle, &mut err_output_handle);
+            }
+        });
+
+        (ok_stream, err_stream)
+    }
+
+    fn map_fallible<D2, E, L>(&self, mut logic: L) -> (Stream<G, D2>, Stream<G, E>)
+    where
+        D2: Data,
+        E: Data,
+        L: FnMut(D1) -> Result<D2, E> + 'static,
+    {
+        let mut storage = Vec::new();
+        self.unary_fallible(Pipeline, "MapFallible", move |_, _| {
+            move |input, ok_output, err_output| {
+                input.for_each(|time, data| {
+                    let mut ok_session = ok_output.session(&time);
+                    let mut err_session = err_output.session(&time);
+                    data.swap(&mut storage);
+                    for d1 in storage.drain(..) {
+                        match logic(d1) {
+                            Ok(d2) => ok_session.give(d2),
+                            Err(e) => err_session.give(e),
+                        }
+                    }
+                })
+            }
+        })
+    }
+
+    fn flat_map_fallible<D2, E, I, L>(&self, mut logic: L) -> (Stream<G, D2>, Stream<G, E>)
+    where
+        D2: Data,
+        E: Data,
+        I: IntoIterator<Item = Result<D2, E>>,
+        L: FnMut(D1) -> I + 'static,
+    {
+        let mut storage = Vec::new();
+        self.unary_fallible(Pipeline, "FlatMapFallible", move |_, _| {
+            move |input, ok_output, err_output| {
+                input.for_each(|time, data| {
+                    let mut ok_session = ok_output.session(&time);
+                    let mut err_session = err_output.session(&time);
+                    data.swap(&mut storage);
+                    for r in storage.drain(..).flat_map(|d1| logic(d1)) {
+                        match r {
+                            Ok(d2) => ok_session.give(d2),
+                            Err(e) => err_session.give(e),
+                        }
+                    }
+                })
+            }
+        })
+    }
+
+    fn filter_fallible<E, L>(&self, mut logic: L) -> (Stream<G, D1>, Stream<G, E>)
+    where
+        E: Data,
+        L: FnMut(&D1) -> Result<bool, E> + 'static,
+    {
+        let mut storage = Vec::new();
+        self.unary_fallible(Pipeline, "FilterFallible", move |_, _| {
+            move |input, ok_output, err_output| {
+                input.for_each(|time, data| {
+                    let mut ok_session = ok_output.session(&time);
+                    let mut err_session = err_output.session(&time);
+                    data.swap(&mut storage);
+                    storage.retain(|d1| match logic(d1) {
+                        Ok(retain) => retain,
+                        Err(e) => {
+                            err_session.give(e);
+                            false
+                        }
+                    });
+                    // TODO(benesch): shouldn't Timely do this check internally?
```

Comment:
Naw, you can send empty buffers around in timely as a way to move resources around. In this case it is 100% fine to guard the `give_vec` the way you have, but if you had a stockpile of empty buffers and you wanted to make them available to the next operator in the sequence, you could repeatedly call `give_vec` to do that.

Reply:
Ack.